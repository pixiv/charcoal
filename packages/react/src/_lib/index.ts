/**
 * 今後ポートされる予定の汎用的な関数群
 */

/**
 * Function used to assert a given code path is unreachable
 */
export function unreachable(): never
/**
 * Function used to assert a given code path is unreachable.
 * Very useful for ensuring switches are exhaustive:
 *
 * ```ts
 * switch (a.type) {
 *   case Types.A:
 *   case Types.B:
 *     break
 *   default:
 *     unreachable(a) // will cause a build error if there was
 *                    // a Types.C that was not checked
 * }
 * ```
 *
 * @param value Value to be asserted as unreachable
 */
// NOTE: Uses separate overloads, _not_ `value?: never`, to not allow `undefined` to be passed

export function unreachable(value: never): never
export function unreachable(value?: never): never {
  throw new Error(
    arguments.length === 0
      ? 'unreachable'
      : `unreachable (${JSON.stringify(value)})`
  )
}

/**
 * 複数のrefをマージする。
 *
 * forwardRefで受け取ったrefと、コンポーネント内で定義したrefを同じ要素につけたいケースなどで使う
 */
export function mergeRefs<T>(...refs: React.Ref<T>[]): React.RefCallback<T> {
  return (value) => {
    for (const ref of refs) {
      if (typeof ref === 'function') {
        ref(value)
      } else if (ref !== null) {
        // TODO: use RefObject
        // https://react.dev/blog/2024/04/25/react-19-upgrade-guide#:~:text=%3D%201%3B-,MutableRef,-is%20now%20deprecated
        ;(ref as React.MutableRefObject<T | null>).current = value
      }
    }
  }
}

export function countCodePointsInString(string: string): number {
  // [...string] とするとproduction buildで動かなくなる
  // cf. https://twitter.com/f_subal/status/1497214727511891972
  return Array.from(string).length
}
