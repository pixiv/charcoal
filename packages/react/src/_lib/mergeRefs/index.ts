import React from 'react'

/**
 * Custom implementation of mergeRefs to replace @react-aria/utils
 * Merges multiple refs into a single ref callback
 */
export function mergeRefs<T>(
  ...refs: (React.Ref<T> | undefined)[]
): React.RefCallback<T> {
  return (value: T | null) => {
    refs.forEach((ref) => {
      if (ref == null) {
        return
      }

      if (typeof ref === 'function') {
        ref(value)
      } else if (typeof ref === 'object') {
        ;(ref as React.MutableRefObject<T | null>).current = value
      }
    })
  }
}
