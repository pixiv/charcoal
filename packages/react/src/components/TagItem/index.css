.charcoal-tag-item {
  --charcoal-tag-item-color: var(--charcoal-text5);
  --charcoal-tag-item-size: 40px;
  --charcoal-tag-item-padding-left: 24px;
  --charcoal-tag-item-padding-right: 24px;

  isolation: isolate;
  position: relative;

  display: inline-flex;
  align-items: center;
  justify-content: center;

  text-decoration: none;
  cursor: pointer;
  overflow: hidden;

  color: var(--charcoal-tag-item-color);
  height: var(--charcoal-tag-item-size);
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: var(--charcoal-tag-item-padding-left);
  padding-right: var(--charcoal-tag-item-padding-right);
  box-sizing: border-box;
  border-radius: 4px;

  transition: 0.2s box-shadow;
}

.charcoal-tag-item[data-size='M'] {
  --charcoal-tag-item-size: 40px;
  --charcoal-tag-item-padding-left: 24px;
  --charcoal-tag-item-padding-right: 24px;
}
.charcoal-tag-item[data-size='S'] {
  --charcoal-tag-item-size: 32px;
  --charcoal-tag-item-padding-left: 16px;
  --charcoal-tag-item-padding-right: 16px;
}
.charcoal-tag-item[data-state='inactive'] {
  --charcoal-tag-item-color: var(--charcoal-text2);
}
.charcoal-tag-item[data-state='active'] {
  --charcoal-tag-item-padding-left: 16px;
  --charcoal-tag-item-padding-right: 8px;
}

.charcoal-tag-item[aria-disabled]:not([aria-disabled='false']) {
  opacity: 0.32;
  cursor: default;
}
.charcoal-tag-item:not([aria-disabled]):focus-visible,
.charcoal-tag-item[aria-disabled='false']:focus-visible {
  box-shadow: 0 0 0 4px rgba(0, 150, 250, 0.32);
}
.charcoal-tag-item:not([aria-disabled]):focus,
.charcoal-tag-item[aria-disabled='false']:focus {
  outline: none;
  box-shadow: 0 0 0 4px rgba(0, 150, 250, 0.32);
}
.charcoal-tag-item:not([aria-disabled]):focus:not(:focus-visible),
.charcoal-tag-item[aria-disabled='false']:focus:not(:focus-visible) {
  box-shadow: none;
}

.charcoal-tag-item__bg {
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
}

.charcoal-tag-item__bg[data-variant='image'] {
  background-color: var(--charcoal-surface4);
}
.charcoal-tag-item__bg[data-variant='image']::before {
  content: '';
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-position: center;
  background-size: cover;
  background-image: var(--charcoal-tag-item-bg);
  mix-blend-mode: overlay;
}

.charcoal-tag-item__bg[data-variant='color'] {
  background-color: var(--charcoal-tag-item-bg);
}
.charcoal-tag-item__bg[data-state='inactive'] {
  background-color: var(--charcoal-surface3);
}

.charcoal-tag-item-text {
  --charcoal-tag-item-text-font-size: 14px;
  --charcoal-tag-item-text-line-height: 22px;
  --charcoal-tag-item-half-leading: calc(
    (
        var(--charcoal-tag-item-text-line-height) -
          var(--charcoal-tag-item-text-font-size)
      ) / 2
  );

  display: flow-root;
  font-size: var(--charcoal-tag-item-text-font-size);
  line-height: var(--charcoal-tag-item-text-line-height);
}

.charcoal-tag-item-text::before {
  display: block;
  width: 0;
  height: 0;
  content: '';
  margin-top: calc(-1 * var(--charcoal-tag-item-half-leading));
}
.charcoal-tag-item-text::after {
  display: block;
  width: 0;
  height: 0;
  content: '';
  margin-bottom: calc(-1 * var(--charcoal-tag-item-half-leading));
}

.charcoal-tag-item__inner {
  display: inline-flex;
  gap: 8px;
  align-items: center;
  z-index: 2;
  height: 100%;
}

.charcoal-tag-item__label-wrapper {
  height: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
}
.charcoal-tag-item__label-wrapper[data-has-translate-label='true'] {
  justify-content: space-between;
}

.charcoal-tag-item__translated-label {
  --charcoal-tag-item-text-font-size: 12px;
  --charcoal-tag-item-text-line-height: 20px;
  font-weight: bold;
}
.charcoal-tag-item__translated-label::before {
  display: none;
}

.charcoal-tag-item__label {
  --charcoal-tag-item-text-font-size: 14px;
  --charcoal-tag-item-text-line-height: 22px;

  max-width: 152px;
  overflow: hidden;
  font-weight: bold;

  color: inherit;
  white-space: nowrap;
  text-overflow: ellipsis;
}
.charcoal-tag-item__label[data-has-translate-label='true'] {
  --charcoal-tag-item-text-font-size: 10px;
  --charcoal-tag-item-text-line-height: 14px;
  font-weight: normal;
}
.charcoal-tag-item__label[data-has-translate-label='true']::after {
  display: none;
}
