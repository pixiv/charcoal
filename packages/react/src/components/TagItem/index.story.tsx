import { action } from '@storybook/addon-actions'
import React from 'react'
import styled from 'styled-components'
import TagItem, { TagItemProps } from '.'
import { Story } from '../../_lib/compat'

export default {
  title: 'TagItem',
  component: TagItem,
  argTypes: {
    color: {
      control: {
        type: 'color',
      },
    },
    bgColor: {
      control: {
        type: 'color',
      },
    },
  },
}

export const Default: Story<TagItemProps> = (props) => {
  return <TagItem {...props} />
}
Default.args = {
  label: '#女の子',
  color: '#F5F5F5',
  bgColor: '#7ACCB1',
}

export const Playground: Story<TagItemProps> = ({ color, bgColor, label, translatedLabel }) => {
  return (
    <div>
      <Container>
        <div>
          <TagItem
            label={label}
            size="M"
            status="default"
            color={color}
            bgColor={bgColor}
            onClick={action('onClick')}
            onClose={action('onClose')}
          />
        </div>
        <div />
        <div>
          <TagItem
            label={label}
            translatedLabel={translatedLabel}
            size="M"
            status="default"
            color={color}
            bgColor={bgColor}
            onClick={action('onClick')}
            onClose={action('onClose')}
          />
        </div>
        <div />
      </Container>
      <Container>
        <div>
          <TagItem
            label={label}
            size="M"
            status="active"
            color={color}
            bgColor={bgColor}
            onClick={action('onClick')}
            onClose={action('onClose')}
          />
        </div>
        <div />
        <div>
          <TagItem
            label={label}
            translatedLabel={translatedLabel}
            size="M"
            status="active"
            color={color}
            bgColor={bgColor}
            onClick={action('onClick')}
            onClose={action('onClose')}
          />
        </div>
        <div />
      </Container>
      <Container>
        <div>
          <TagItem
            label={label}
            size="M"
            status="inactive"
            color={color}
            bgColor={bgColor}
            onClick={action('onClick')}
            onClose={action('onClose')}
          />
        </div>
        <div>
          <TagItem
            label={label}
            size="M"
            status="inactive"
            disabled
            color={color}
            bgColor={bgColor}
            onClick={action('onClick')}
            onClose={action('onClose')}
          />
        </div>
        <div>
          <TagItem
            label={label}
            translatedLabel={translatedLabel}
            size="M"
            status="inactive"
            color={color}
            bgColor={bgColor}
            onClick={action('onClick')}
            onClose={action('onClose')}
          />
        </div>
        <div>
          <TagItem
            label={label}
            translatedLabel={translatedLabel}
            size="M"
            status="inactive"
            disabled
            color={color}
            bgColor={bgColor}
            onClick={action('onClick')}
            onClose={action('onClose')}
          />
        </div>
      </Container>
      <Container>
        <div>
          <TagItem
            label={label}
            size="S"
            status="default"
            color={color}
            bgColor={bgColor}
            onClick={action('onClick')}
            onClose={action('onClose')}
          />
        </div>
      </Container>
      <Container>
        <div>
          <TagItem
            label={label}
            size="S"
            status="active"
            color={color}
            bgColor={bgColor}
            onClick={action('onClick')}
            onClose={action('onClose')}
          />
        </div>
      </Container>
      <Container>
        <div>
          <TagItem
            label={label}
            size="S"
            status="inactive"
            color={color}
            bgColor={bgColor}
            onClick={action('onClick')}
            onClose={action('onClose')}
          />
        </div>
        <div>
          <TagItem
            label={label}
            size="S"
            status="inactive"
            disabled
            color={color}
            bgColor={bgColor}
            onClick={action('onClick')}
            onClose={action('onClose')}
          />
        </div>
      </Container>
    </div>
  )
}

const Container = styled.div`
  display: grid;
  grid-template-columns: repeat(4, 1fr);

  & > div {
    display: flex;
    align-items: center;
    justify-content: center;
    padding: 1em;
  }
`

Playground.args = {
  label: '#女の子',
  translatedLabel: "girl",
  color: '#F5F5F5',
  bgColor: '#7ACCB1',
}
