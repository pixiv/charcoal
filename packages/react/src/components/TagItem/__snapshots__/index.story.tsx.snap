// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots TagItem Default 1`] = `
.c0 {
  isolation: isolate;
  position: relative;
  height: 40px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  overflow: hidden;
  border-radius: 4px;
  padding-right: 24px;
  padding-left: 24px;
  color: var(--charcoal-text5);
  -webkit-transition: 0.2s box-shadow;
  transition: 0.2s box-shadow;
}

.c0:not(:disabled):not([aria-disabled]):focus,
.c0[aria-disabled='false']:focus,
.c0:not(:disabled):not([aria-disabled]):active,
.c0[aria-disabled='false']:active,
.c0:not(:disabled):not([aria-disabled]):focus-visible,
.c0[aria-disabled='false']:focus-visible {
  outline: none;
  box-shadow: 0 0 0 4px rgba(0,150,250,0.32);
}

.c0:disabled,
.c0[aria-disabled]:not([aria-disabled='false']) {
  opacity: 0.32;
  cursor: default;
}

.c1 {
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: #7ACCB1;
}

.c2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  gap: 8px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  z-index: 2;
}

.c3 {
  font-size: 14px;
  line-height: 22px;
  font-weight: bold;
  display: flow-root;
}

.c3::before {
  display: block;
  width: 0;
  height: 0;
  content: '';
  margin-top: -4px;
}

.c3::after {
  display: block;
  width: 0;
  height: 0;
  content: '';
  margin-bottom: -4px;
}

.c4 {
  max-width: 152px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  font-size: inherit;
  color: inherit;
  line-height: inherit;
}

<div
  data-dark={false}
>
  <a
    className="c0"
    href=""
    onClick={[Function]}
    onDragStart={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchCancel={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    rel=""
    role="button"
    size="M"
    tabIndex={0}
    target=""
  >
    <div
      className="c1"
    />
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <span
          className="c4"
        >
          #女の子
        </span>
      </div>
    </div>
  </a>
</div>
`;

exports[`Storyshots TagItem Playground 1`] = `
.c1 {
  isolation: isolate;
  position: relative;
  height: 40px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  overflow: hidden;
  border-radius: 4px;
  padding-right: 24px;
  padding-left: 24px;
  color: var(--charcoal-text5);
  -webkit-transition: 0.2s box-shadow;
  transition: 0.2s box-shadow;
}

.c1:not(:disabled):not([aria-disabled]):focus,
.c1[aria-disabled='false']:focus,
.c1:not(:disabled):not([aria-disabled]):active,
.c1[aria-disabled='false']:active,
.c1:not(:disabled):not([aria-disabled]):focus-visible,
.c1[aria-disabled='false']:focus-visible {
  outline: none;
  box-shadow: 0 0 0 4px rgba(0,150,250,0.32);
}

.c1:disabled,
.c1[aria-disabled]:not([aria-disabled='false']) {
  opacity: 0.32;
  cursor: default;
}

.c8 {
  isolation: isolate;
  position: relative;
  height: 40px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  overflow: hidden;
  border-radius: 4px;
  padding-right: 8px;
  padding-left: 16px;
  color: var(--charcoal-text5);
  -webkit-transition: 0.2s box-shadow;
  transition: 0.2s box-shadow;
}

.c8:not(:disabled):not([aria-disabled]):focus,
.c8[aria-disabled='false']:focus,
.c8:not(:disabled):not([aria-disabled]):active,
.c8[aria-disabled='false']:active,
.c8:not(:disabled):not([aria-disabled]):focus-visible,
.c8[aria-disabled='false']:focus-visible {
  outline: none;
  box-shadow: 0 0 0 4px rgba(0,150,250,0.32);
}

.c8:disabled,
.c8[aria-disabled]:not([aria-disabled='false']) {
  opacity: 0.32;
  cursor: default;
}

.c9 {
  isolation: isolate;
  position: relative;
  height: 40px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  overflow: hidden;
  border-radius: 4px;
  padding-right: 24px;
  padding-left: 24px;
  color: var(--charcoal-text2);
  -webkit-transition: 0.2s box-shadow;
  transition: 0.2s box-shadow;
}

.c9:not(:disabled):not([aria-disabled]):focus,
.c9[aria-disabled='false']:focus,
.c9:not(:disabled):not([aria-disabled]):active,
.c9[aria-disabled='false']:active,
.c9:not(:disabled):not([aria-disabled]):focus-visible,
.c9[aria-disabled='false']:focus-visible {
  outline: none;
  box-shadow: 0 0 0 4px rgba(0,150,250,0.32);
}

.c9:disabled,
.c9[aria-disabled]:not([aria-disabled='false']) {
  opacity: 0.32;
  cursor: default;
}

.c11 {
  isolation: isolate;
  position: relative;
  height: 32px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  overflow: hidden;
  border-radius: 4px;
  padding-right: 16px;
  padding-left: 16px;
  color: var(--charcoal-text5);
  -webkit-transition: 0.2s box-shadow;
  transition: 0.2s box-shadow;
}

.c11:not(:disabled):not([aria-disabled]):focus,
.c11[aria-disabled='false']:focus,
.c11:not(:disabled):not([aria-disabled]):active,
.c11[aria-disabled='false']:active,
.c11:not(:disabled):not([aria-disabled]):focus-visible,
.c11[aria-disabled='false']:focus-visible {
  outline: none;
  box-shadow: 0 0 0 4px rgba(0,150,250,0.32);
}

.c11:disabled,
.c11[aria-disabled]:not([aria-disabled='false']) {
  opacity: 0.32;
  cursor: default;
}

.c12 {
  isolation: isolate;
  position: relative;
  height: 32px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  overflow: hidden;
  border-radius: 4px;
  padding-right: 8px;
  padding-left: 16px;
  color: var(--charcoal-text5);
  -webkit-transition: 0.2s box-shadow;
  transition: 0.2s box-shadow;
}

.c12:not(:disabled):not([aria-disabled]):focus,
.c12[aria-disabled='false']:focus,
.c12:not(:disabled):not([aria-disabled]):active,
.c12[aria-disabled='false']:active,
.c12:not(:disabled):not([aria-disabled]):focus-visible,
.c12[aria-disabled='false']:focus-visible {
  outline: none;
  box-shadow: 0 0 0 4px rgba(0,150,250,0.32);
}

.c12:disabled,
.c12[aria-disabled]:not([aria-disabled='false']) {
  opacity: 0.32;
  cursor: default;
}

.c13 {
  isolation: isolate;
  position: relative;
  height: 32px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  overflow: hidden;
  border-radius: 4px;
  padding-right: 16px;
  padding-left: 16px;
  color: var(--charcoal-text2);
  -webkit-transition: 0.2s box-shadow;
  transition: 0.2s box-shadow;
}

.c13:not(:disabled):not([aria-disabled]):focus,
.c13[aria-disabled='false']:focus,
.c13:not(:disabled):not([aria-disabled]):active,
.c13[aria-disabled='false']:active,
.c13:not(:disabled):not([aria-disabled]):focus-visible,
.c13[aria-disabled='false']:focus-visible {
  outline: none;
  box-shadow: 0 0 0 4px rgba(0,150,250,0.32);
}

.c13:disabled,
.c13[aria-disabled]:not([aria-disabled='false']) {
  opacity: 0.32;
  cursor: default;
}

.c2 {
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: #7ACCB1;
}

.c10 {
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: #7ACCB1;
  background-color: var(--charcoal-surface3);
}

.c3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  gap: 8px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  z-index: 2;
}

.c4 {
  font-size: 14px;
  line-height: 22px;
  font-weight: bold;
  display: flow-root;
}

.c4::before {
  display: block;
  width: 0;
  height: 0;
  content: '';
  margin-top: -4px;
}

.c4::after {
  display: block;
  width: 0;
  height: 0;
  content: '';
  margin-bottom: -4px;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-size: 10px;
}

.c5 {
  max-width: 152px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  font-size: inherit;
  color: inherit;
  line-height: inherit;
}

.c7 {
  font-size: 12px;
  line-height: 20px;
  font-weight: bold;
  display: flow-root;
}

.c7::before {
  display: block;
  width: 0;
  height: 0;
  content: '';
  margin-top: -4px;
}

.c7::after {
  display: block;
  width: 0;
  height: 0;
  content: '';
  margin-bottom: -4px;
}

.c0 {
  display: grid;
  grid-template-columns: repeat(4,1fr);
}

.c0 > div {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 1em;
}

<div
  data-dark={false}
>
  <div>
    <div
      className="c0"
    >
      <div>
        <a
          className="c1"
          onClick={[Function]}
          onDragStart={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          size="M"
          tabIndex={0}
        >
          <div
            className="c2"
          />
          <div
            className="c3"
          >
            <div
              className="c4"
            >
              <span
                className="c5"
              >
                #女の子
              </span>
            </div>
          </div>
        </a>
      </div>
      <div />
      <div>
        <a
          className="c1"
          onClick={[Function]}
          onDragStart={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          size="M"
          tabIndex={0}
        >
          <div
            className="c2"
          />
          <div
            className="c3"
          >
            <div
              className="c6"
            >
              <div
                className="c7"
              >
                <span
                  className="c5"
                >
                  girl
                </span>
              </div>
              <span
                className="c5"
              >
                #女の子
              </span>
            </div>
          </div>
        </a>
      </div>
      <div />
    </div>
    <div
      className="c0"
    >
      <div>
        <a
          className="c8"
          onClick={[Function]}
          onDragStart={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          size="M"
          tabIndex={0}
        >
          <div
            className="c2"
          />
          <div
            className="c3"
          >
            <div
              className="c4"
            >
              <span
                className="c5"
              >
                #女の子
              </span>
            </div>
            <pixiv-icon
              name="16/Remove"
            />
          </div>
        </a>
      </div>
      <div />
      <div>
        <a
          className="c8"
          onClick={[Function]}
          onDragStart={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          size="M"
          tabIndex={0}
        >
          <div
            className="c2"
          />
          <div
            className="c3"
          >
            <div
              className="c6"
            >
              <div
                className="c7"
              >
                <span
                  className="c5"
                >
                  girl
                </span>
              </div>
              <span
                className="c5"
              >
                #女の子
              </span>
            </div>
            <pixiv-icon
              name="16/Remove"
            />
          </div>
        </a>
      </div>
      <div />
    </div>
    <div
      className="c0"
    >
      <div>
        <a
          className="c9"
          onClick={[Function]}
          onDragStart={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          size="M"
          tabIndex={0}
        >
          <div
            className="c10"
          />
          <div
            className="c3"
          >
            <div
              className="c4"
            >
              <span
                className="c5"
              >
                #女の子
              </span>
            </div>
          </div>
        </a>
      </div>
      <div>
        <a
          aria-disabled={true}
          className="c9"
          onClick={[Function]}
          onDragStart={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          size="M"
        >
          <div
            className="c10"
          />
          <div
            className="c3"
          >
            <div
              className="c4"
            >
              <span
                className="c5"
              >
                #女の子
              </span>
            </div>
          </div>
        </a>
      </div>
      <div>
        <a
          className="c9"
          onClick={[Function]}
          onDragStart={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          size="M"
          tabIndex={0}
        >
          <div
            className="c10"
          />
          <div
            className="c3"
          >
            <div
              className="c6"
            >
              <div
                className="c7"
              >
                <span
                  className="c5"
                >
                  girl
                </span>
              </div>
              <span
                className="c5"
              >
                #女の子
              </span>
            </div>
          </div>
        </a>
      </div>
      <div>
        <a
          aria-disabled={true}
          className="c9"
          onClick={[Function]}
          onDragStart={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          size="M"
        >
          <div
            className="c10"
          />
          <div
            className="c3"
          >
            <div
              className="c6"
            >
              <div
                className="c7"
              >
                <span
                  className="c5"
                >
                  girl
                </span>
              </div>
              <span
                className="c5"
              >
                #女の子
              </span>
            </div>
          </div>
        </a>
      </div>
    </div>
    <div
      className="c0"
    >
      <div>
        <a
          className="c11"
          onClick={[Function]}
          onDragStart={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          size="S"
          tabIndex={0}
        >
          <div
            className="c2"
          />
          <div
            className="c3"
          >
            <div
              className="c4"
            >
              <span
                className="c5"
              >
                #女の子
              </span>
            </div>
          </div>
        </a>
      </div>
    </div>
    <div
      className="c0"
    >
      <div>
        <a
          className="c12"
          onClick={[Function]}
          onDragStart={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          size="S"
          tabIndex={0}
        >
          <div
            className="c2"
          />
          <div
            className="c3"
          >
            <div
              className="c4"
            >
              <span
                className="c5"
              >
                #女の子
              </span>
            </div>
            <pixiv-icon
              name="16/Remove"
            />
          </div>
        </a>
      </div>
    </div>
    <div
      className="c0"
    >
      <div>
        <a
          className="c13"
          onClick={[Function]}
          onDragStart={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          size="S"
          tabIndex={0}
        >
          <div
            className="c10"
          />
          <div
            className="c3"
          >
            <div
              className="c4"
            >
              <span
                className="c5"
              >
                #女の子
              </span>
            </div>
          </div>
        </a>
      </div>
      <div>
        <a
          aria-disabled={true}
          className="c13"
          onClick={[Function]}
          onDragStart={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          size="S"
        >
          <div
            className="c10"
          />
          <div
            className="c3"
          >
            <div
              className="c4"
            >
              <span
                className="c5"
              >
                #女の子
              </span>
            </div>
          </div>
        </a>
      </div>
    </div>
  </div>
</div>
`;
